infra/concourse/build/developer-tools/build/scripts/setup_environment.sh

```diff
@@ -0,0 +1,64 @@
+#! /bin/bash
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+if [[ -z "${SERVICE_ACCOUNT_JSON}" ]]; then
+  echo "Error: SERVICE_ACCOUNT_JSON must contain the JSON string (not the" >&2
+  echo "file path) of the service account required to execute " >&2
+  echo "Terraform/gsutil/gcloud. For example: " >&2
+  echo "export SERVICE_ACCOUNT_JSON=\$(< /path/to/credentials.json)" >&2
+  exit 1
+fi
+
+# Always clean up.
+DELETE_AT_EXIT="$(mktemp -d)"
+
+finish() {
+  [[ -d "${DELETE_AT_EXIT}" ]] && rm -rf "${DELETE_AT_EXIT}"
+}
+
+setup_environment() {
+  local tmpfile
+  tmpfile="$(mktemp)"
+  echo "${SERVICE_ACCOUNT_JSON}" > "${tmpfile}"
+
+  # Terraform and most other tools respect GOOGLE_CREDENTIALS
+  # https://www.terraform.io/docs/providers/google/provider_reference.html#credentials-1
+  export GOOGLE_CREDENTIALS="${SERVICE_ACCOUNT_JSON}"
+
+  # gcloud variables
+  export CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE="${tmpfile}"
+
+  # InSpec respects GOOGLE_APPLICATION_CREDENTIALS
+  # https://github.com/inspec/inspec-gcp#create-credentials-file-via
+  export GOOGLE_APPLICATION_CREDENTIALS="${tmpfile}"
+
+  # Configure gsutil standalone
+  # https://cloud.google.com/storage/docs/gsutil/commands/config
+  gcloud config set pass_credentials_to_gsutil false
+  echo "[Credentials]" > ~/.boto
+  echo "gs_service_key_file = ${tmpfile}" >> ~/.boto
+}
+
+main() {
+  set -eu
+  # Setup trap handler to auto-cleanup
+  export TMPDIR="${DELETE_AT_EXIT}"
+  trap finish EXIT
+
+  # Setup environment
+  setup_environment
+}
+
+main "$@"
```

Comment:
This script is conflating up two separate modes of operation from `ci_integration.sh`.  This causes a problem because `set -eu` is enabled in the interactive shell.

The two modes are:

 1. The script is sourced from an interactive shell, e.g. `source setup_environment.sh`.
 2. The script is executed directly, e.g. `./setup_environment.sh`.

Mode 1 is intended for a "library of functions", e.g. "load the library, then setup the environment via `setup_environment()`.

Mode 2 is intended to execute a task, e.g. `"Execute the test lifecycle` here: https://github.com/terraform-google-modules/terraform-google-project-factory/blob/f65c1402fb84932b7a933f50883fdc97dff8d2ee/test/ci_integration.sh#L65

Mode 2 should have the trap handler because the main task should clean up after itself..  Mode 1 should not have a trap handler because when used as a library, the parent shell is responsible for cleanup.

Suggested changes:
 * [ ] Add `if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then ...` as per https://github.com/terraform-google-modules/terraform-google-project-factory/blob/f65c1402fb84932b7a933f50883fdc97dff8d2ee/test/ci_integration.sh#L72
 * [ ] Implement main()'s behavior as simply executing some command with the correct environment.  So main() sets up the trap handler, sets up the environment, then executes "$@"
 * [ ] Make sure `set -e and set -u` are only enabled in mode 2.

Reply:
IMO we should go for the even simpler approach of keeping this script as a "library of functions" and either:
- Always require that one of the functions be explicitly invoked
- Or, have a separate script which is meant to be directly invoked (and can source this script)

Reply:
I've updated the script with the conditional logic, but we can split it out if that's preferred

Reply:
Resolved with the most recent commit